Fixes #4195 : Added dark mode support to QuestionPlayerActivity and ExplorationActivity

bhaktideshmukh opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation

Fixes #4195 .
Added dark mode support to QuestionPlayerActivity and ExplorationActivity.

Mocks Links

- QuestionPlayer : https://xd.adobe.com/view/c05e9343-60f6-4c11-84ac-c756b75b940f-950d/screen/8fbb663a-39aa-4be1-83bd-73bde6f85dcc/specs/
- ExplorationActivity : https://xd.adobe.com/view/c05e9343-60f6-4c11-84ac-c756b75b940f-950d/screen/40d6cf5f-f673-42c8-af15-035f711ef6b9/specs/

**ScreenShots**

**_QuestionPlayerActivity_**

| Before  | After dark mode|
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157729236-3b585bf2-0f0b-4d33-9788-7c8bdfa6630e.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158637806-4912a20e-855b-438c-b9f8-28caf673ffd6.jpeg" width="250" />|

| Before  | After dark mode |
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157729286-8b5b7c8e-9257-4095-aa8d-f6ef27c6eee2.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158637973-e6d727ee-a88f-4607-9dd4-f7e8928bd8cd.jpeg" width="250" />|

| Before  | After dark mode |
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157729602-18003f0a-e7dc-4cba-8b16-ff19b5b400e6.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158637985-0c38cee6-9fa6-4abf-ab47-42eee2ef9347.jpeg" width="250" />|

**_ExplorationActivity_**

| Before  | After dark mode |
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157729989-724a214e-407d-45ce-aa77-b8025c8a2bf4.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158638541-88c2e0ee-907f-457e-b40a-2d33d752cff9.jpeg" width="250" />|

| Before  | After dark mode |
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157730034-4fd0d5fd-1b8a-4fa8-97a6-e3a1ed7249bd.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158638713-a3c01255-5dbb-45ae-a0b5-5a1e712f81f9.jpeg" width="250" />|

| Before  | After dark mode |
|-------|---|
|<img src="https://user-images.githubusercontent.com/78796264/157730063-7a327312-18e4-4d9f-80c3-d8e909189182.jpg" width="250" />|<img src="https://user-images.githubusercontent.com/78796264/158638698-54c055af-39ef-4b3e-89bc-05bec332305a.jpeg" width="250" />|

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

bhaktideshmukh committed:
added dark mode

bhaktideshmukh committed:
Merge branch 'develop' into dark_mode

bhaktideshmukh committed:
added color_def

rt4914 commented:
Un-assigning until approved / re-assigned by @ayush0402

bhaktideshmukh committed:
updated toolbar color

bhaktideshmukh commented:
PTAL @ayush0402 
Thanks

ayush0402 commented on app/src/main/res/layout-land/question_player_fragment.xml:
```diff
@@ -13,7 +13,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:background="@color/question_player_background">
+    android:background="@color/question_player_activity_layout_background_color">
```
Please update all colors to follow newer convention having file name in the color declaration.

bhaktideshmukh replied:
> Please update all colors to follow newer convention having file name in the color declaration.

Can you please elaborate how should I update the name here.
I have already added the file name in the beginning of the color name.

ayush0402 commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -35,5 +39,12 @@
   <!-- Pin Password Activity -->
+  <color name="pin_password_activity_layout_background_color">@color/color_palette_background_color</color>
   <color name="component_color_pin_password_activity_layout_background_color">@color/color_palette_background_color</color>
+  <!-- QuestionPlayerActivity -->
```
`<!-- Question Player Activity -->`

bhaktideshmukh replied:
Done

ayush0402 commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -35,5 +39,12 @@
+  <color name="question_player_activity_layout_background_color">@color/color_palette_primary_background_color</color>
+  <color name="question_player_activity_textview_highlighted_color">@color/color_palette_highlighted_text_color</color>
+  <color name="question_player_activity_textview_secondary_color">@color/color_palette_secondary_text_color</color>
+  <!-- ExplorationActivity -->
```
`<!-- Exploration Activity -->`

bhaktideshmukh replied:
Done

ayush0402 commented on app/src/main/res/layout/question_player_activity.xml:
```diff
@@ -20,7 +21,7 @@
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
         android:layout_gravity="end"
-        android:background="@color/oppia_primary"
+        android:background="@color/color_def_oppia_green"
```
This should be a color from component_colors.xml

bhaktideshmukh replied:
Done

ayush0402 reviewed (changes requested):
The approach seems to be good. Took a pass and left some comments. Please update the name convention and assign me back. Rest seems to be fine might need to rename some colors later.

bhaktideshmukh committed:
added toolbar_color to component_colors

rt4914 commented:
@ayush0402 Can you review it once again as you have suggested changes.

rt4914 commented on app/src/main/res/layout-land/question_player_fragment.xml:
```diff
@@ -13,7 +13,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:background="@color/question_player_background">
+    android:background="@color/question_player_activity_layout_background_color">
```
For this create a new color `question_player_fragment_background_color` as this is a fragment file and not an activity file. Also, removed `layout` word as this does not add any extra info.

bhaktideshmukh replied:
Done

rt4914 commented on app/src/main/res/layout-land/question_player_fragment.xml:
```diff
@@ -28,7 +28,7 @@
         android:fontFamily="sans-serif"
         android:text="@string/question_training_session_finished_header"
-        android:textColor="@color/oppia_primary_text"
+        android:textColor="@color/question_player_activity_textview_secondary_color"
```
Same -> `question_player_fragment_...`

bhaktideshmukh replied:
Done

rt4914 commented on app/src/main/res/layout-land/question_player_fragment.xml:
```diff
@@ -43,7 +43,7 @@
         android:fontFamily="sans-serif"
         android:text="@string/question_training_session_finished_message"
-        android:textColor="@color/oppia_primary_text"
+        android:textColor="@color/question_player_activity_textview_secondary_color"
```
Ditto

rt4914 commented on app/src/main/res/layout/content_item.xml:
```diff
@@ -48,7 +48,7 @@
       android:fontFamily="sans-serif"
       android:text="@{htmlContent}"
-      android:textColor="@color/oppia_primary_text"
+      android:textColor="@color/question_player_activity_textview_highlighted_color"
```
Considering this color might be getting used at a lot of places, I suggest you name it as follows:

`shared_player_text_view_primary_color`

rt4914 replied:
Or if you want to create specific colors then name it as `content_item_text_view_color`

bhaktideshmukh replied:
To me first one looks much better

rt4914 commented on app/src/main/res/layout/exploration_fragment.xml:
```diff
@@ -10,6 +10,7 @@
       android:id="@+id/state_fragment_placeholder"
       android:layout_width="match_parent"
       android:layout_height="match_parent"
+      android:background="@color/exploration_activity_layout_background_color"
```
Replace -> `activity` -> `fragment`


Also if all background colors are same then you can also create a shared one like `shared_player_background_color`

bhaktideshmukh replied:
Done

rt4914 commented on app/src/main/res/layout/feedback_item.xml:
```diff
@@ -48,7 +48,7 @@
       android:fontFamily="sans-serif"
       android:text="@{htmlContent}"
-      android:textColor="@color/oppia_primary_text"
+      android:textColor="@color/question_player_activity_textview_highlighted_color"
```
This naming is incorrect because this file is getting used in exploration player too.

Name something like 

`feedback_item_text_view_primary_color`

rt4914 replied:
Ditto for other files too.

rt4914 commented on app/src/main/res/layout/previous_responses_header_item.xml:
```diff
@@ -42,7 +42,7 @@
       android:layout_width="0dp"
       android:layout_height="2dp"
-      android:background="@color/color_def_mid_grey"
+      android:background="@color/question_player_activity_textview_highlighted_color"
```
Ditto for this file

rt4914 commented on app/src/main/res/layout/question_player_fragment.xml:
```diff
@@ -15,7 +15,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:background="@color/question_player_background">
+    android:background="@color/question_player_activity_layout_background_color">
```
`activity` -> `fragment` replace it.

bhaktideshmukh replied:
Done

rt4914 reviewed (changes requested):
@bhaktideshmukh Suggested changes. Thanks.

rt4914 commented:
@bhaktideshmukh I think all colors should have `component_color...` keyword ahead of it but I am unable see that rule in wiki https://github.com/oppia/oppia-android/wiki/Dark-Mode

Will confirm this with Ayush.


@bhaktideshmukh DON'T apply the changes that I have suggested right now as I have some confusion with the rules and therefore I think let's first clarify them. Also at a lot of places I have suggested `fragment` instead of `activity` which might not be correct as per wiki.

bhaktideshmukh committed:
changes naming conventions

bhaktideshmukh committed:
Merge branch 'develop' into dark_mode

rt4914 reviewed (changes requested):
@bhaktideshmukh Please add dark-mode mock links in PR description and also apply the suggested changes.

bhaktideshmukh commented:
> @bhaktideshmukh Please add dark-mode mock links in PR description and also apply the suggested changes.

I have added some screenshots in PR description. Do I have to add the screenshots of all the pages for dark mode of QuestionPlayerActivity and ExplorationActivity in a doc and then attach the link here?
And about suggested changes I have already added all the suggested changes.

rt4914 commented on app/src/main/res/layout-land/question_player_fragment.xml:
```diff
@@ -13,7 +13,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:background="@color/question_player_background">
+    android:background="@color/question_player_fragment_background_color">
```
All colors should follow this format:

`component_color_<activity-name>_<color-usage/type>_color`

For this it should be `component_color_question_player_activity_background_color`

bhaktideshmukh replied:
Got it

rt4914 commented on app/src/main/res/values/color_palette.xml:
```diff
@@ -16,4 +16,9 @@
+  <color name="color_palette_secondary_text_color">@color/color_def_oppia_stroke_black</color>
+  <color name="color_palette_primary_background_color">@color/color_def_white_light</color>
+  <color name="color_palette_dialog_primary_text_color">@color/color_def_black</color>
+
```
Remove this blank line

rt4914 reviewed (changes requested):
Now sure why in last review the comments were not shown but basically follow the convention mentioned in the below comment.

bhaktideshmukh committed:
updated color names

bhaktideshmukh committed:
Merge remote-tracking branch 'upstream/develop' into dark_mode

Akshatkamboj14 reviewed (changes requested):
1. I think exploration activity's card background is not correct , it should be **#32363B** .
2. Question Player Activity's toolbar color should be changed too, in the Dark Mode mocks its mentioned **#035D63**.
3. Kindly PTAL at the buttons, it should also have borders, like this - 
![image](https://user-images.githubusercontent.com/92685493/168970937-6a1fe916-cc2f-45a3-864b-3c8a92066c5a.png)
![image](https://user-images.githubusercontent.com/92685493/168971030-ba99d102-86be-4af5-86cd-5ea30709913a.png)
and background color of the continue button have changes too.

4. the progress bar color also have changes left, it should be like this
![image](https://user-images.githubusercontent.com/92685493/168971858-9faebca7-6c90-4108-a5be-269264922d05.png)

rt4914 commented:
For now we have finalised a different approach, which is we will implement dark-mode for items like content, feedback, interactions, buttons, first and then in the end focus on fragment + activity.

Therefore this PR is on hold for now.

bhaktideshmukh commented:
As per decided will be doing this PR in subparts starting from PR no. https://github.com/oppia/oppia-android/pull/4377 hence closing this one.
